tool/mesos/singa_scheduler.cc

```diff
@@ -0,0 +1,439 @@
+* or more contributor license agreements.  See the NOTICE file
+* distributed with this work for additional information
+* regarding copyright ownership.  The ASF licenses this file
+* to you under the Apache License, Version 2.0 (the
+* "License"); you may not use this file except in compliance
+* with the License.  You may obtain a copy of the License at
+* 
+*   http://www.apache.org/licenses/LICENSE-2.0
+* 
+* Unless required by applicable law or agreed to in writing,
+* software distributed under the License is distributed on an
+* "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+* KIND, either express or implied.  See the License for the
+* specific language governing permissions and limitations
+* under the License.
+*
+*************************************************************/
+
+#include <stdio.h>
+#include <glog/logging.h>
+#include <stdlib.h>
+#include <mesos/scheduler.hpp>
+#include <string>
+#include <iostream>
+#include <fstream>
+#include <hdfs.h>
+#include "singa/proto/job.pb.h"
+#include "singa/proto/singa.pb.h"
+#include "scheduler.pb.h"
+#include "singa/utils/common.h"
+#include <google/protobuf/text_format.h>
+
+/**
+ * \file singa_scheduler.cc implements a framework for managing SINGA jobs.
+ *  
+ * The scheduler takes a job configuration file [file] and performs the following:
+ *	1. Parse the config file to determine the required resources. 
+ *	2. [Optional] Copy singa.conf to the HDFS: /singa/singa.conf.
+ *			2.1. Raise error if singa.conf is NOT FOUND on HDFS
+ *	3. Wait for offers from the Mesos master until enough is acquired. 
+ *	4. Keep an increasing counter of job ID.  
+ *	5. Write [file] to HDFS: /singa/job_ID/job.conf
+ *	6. Start the task:
+ *			+ Set URI in the TaskInfo message to point to the config files on HDFS:
+							/singa/singa.conf
+							/singa/Job_ID/job.conf
+ *			+ Set the executable command: singa-run.sh -conf ./job.conf
+ *
+ *	We assume that singa-run.sh is include in $PATH variable at all nodes. 
+ *	(Else, we set the executable to its full path)
+ *	./job.conf is relative path pointing to the current sandbox directory created
+ *	dynamically by Mesos. 
+ *
+ *
+ * Scheduling: 
+ *	Each SINGA job requires certain resources represented by: (1) number of workers, (2) number of worker groups
+ *	and (3) number of worker per process. The resources offered by Mesos contains (1) number of host, (2) number of CPUs
+ *	at each host and (3) memory available at each hosts. 
+ *
+ *	Our scheduler performs simply task assignment which guarantees that each process runs an entire work group,
+ *	and each takes all the memory offered by the slave. We assume that each slave runs ONE process, that is the following
+ *	condition holds:
+ *		nCPUs_per_host >= nWorkersPerProcess+nServersPerProcess //if seperate
+										 >= max(nWorkerPerProcess, nServersPerProcess) // if attached
+ */
+using std::string;
+using mesos::SchedulerDriver;
+using std::vector;
+using std::map;
```

Comment:
please check the format using
    
    python tool/cpplint.py tool/mesos/*.cc
    python tool/cpplint.py tool/mesos/*.h